File: tests/library.test.ts
import assert from 'node:assert/strict';
import { mkdir, mkdtemp, readFile, readdir, rm, symlink, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { dirname, join } from 'node:path';
import { test } from 'node:test';
import {
  bundledProfiles,
  checkDraft,
  defaultUnitProfile,
  draftUnit,
  prepareRequest,
  reviewDraft,
} from '../src/core/index.js';
import type { LibraryEntry, LibraryIconsResponse, LibraryState } from '../src/lab/contracts.js';
import { LabLibrary } from '../src/lab/library.js';
import { startLab } from '../src/lab/server.js';
import { FakeModel, miraCandidate, miraRequest } from './fixtures/core-fixtures.js';

async function fixture() {
  const root = await mkdtemp(join(tmpdir(), 'unitlab-library-'));
  const options = {
    settingsFile: join(root, 'settings.json'),
    defaultDirectory: join(root, 'library'),
  };
  const library = await LabLibrary.open(options);
  return { root, options, library, cleanup: () => rm(root, { recursive: true, force: true }) };
}

test('library saves every stage unchanged, deduplicates and restores across server lifetimes', async () => {
  const { library, options, cleanup } = await fixture();
  try {
    assert.deepEqual(await library.state(), { directory: options.defaultDirectory, entries: [] });
    const model = new FakeModel();
    const prepared = await prepareRequest(miraRequest());
    const draft = await draftUnit(prepared, model);
    const checked = await checkDraft(draft);
    const result = await reviewDraft(checked, model);
    for (const artifact of [prepared, draft, checked, result]) {
      const before = structuredClone(artifact);
      const [first, second] = await Promise.all([library.save(artifact), library.save(artifact)]);
      assert.deepEqual(first, second);
      assert.deepEqual(first.character, miraRequest().character);
      assert.equal(first.kind, artifact.kind);
      assert.deepEqual((await library.load(first.id)).artifact, before);
      assert.deepEqual(artifact, before);
    }
    const reopened = await LabLibrary.open(options);
    assert.deepEqual(await reopened.state(), await library.state());
    assert.equal((await library.state()).entries.length, 4);
    assert.equal((await readdir(options.defaultDirectory)).length, 4);
    await assert.rejects(library.save({ ...prepared, apiKey: 'must-not-save' }));
    const tampered = structuredClone(prepared);
    tampered.request.task = 'Changed without preparation.';
    await assert.rejects(library.save(tampered), /hash/);
    await assert.rejects(library.save(miraRequest()), /completed stage/);
  } finally {
    await cleanup();
  }
});

test('changing the folder persists the setting and leaves the old library intact', async () => {
  const { root, library, options, cleanup } = await fixture();
  try {
    const prepared = await prepareRequest(miraRequest());
    const saved = await library.save(prepared);
    const alternate = join(root, 'chosen', 'units');
    assert.deepEqual(await library.configure(alternate), { directory: alternate, entries: [] });
    assert.deepEqual(await (await LabLibrary.open(options)).state(), {
      directory: alternate,
      entries: [],
    });
    assert.ok(await readFile(join(options.defaultDirectory, `unitlab-${saved.id}.json`)));
    await library.save(prepared);
    assert.equal((await library.configure(options.defaultDirectory)).entries.length, 1);
    const settings = JSON.parse(await readFile(options.settingsFile, 'utf8'));
    assert.deepEqual(Object.keys(settings).sort(), ['directory', 'unitLabSettings']);
    await assert.rejects(library.configure('  '));
    assert.equal((await library.state()).directory, options.defaultDirectory);
  } finally {
    await cleanup();
  }
});

test('deletion accepts only intact managed documents and preserves other files and folders', async () => {
  const { root, library, options, cleanup } = await fixture();
  try {
    const saved = await library.save(await prepareRequest(miraRequest()));
    const ordinary = join(options.defaultDirectory, 'notes.json');
    await writeFile(ordinary, '{"important":"keep"}');
    const unknownId = '0'.repeat(64);
    const unknown = join(options.defaultDirectory, `unitlab-${unknownId}.json`);
    await writeFile(unknown, '{"unrelated":true}');
    const linkedId = '1'.repeat(64);
    const external = join(root, 'external.json');
    await writeFile(external, 'external content');
    const linked = join(options.defaultDirectory, `unitlab-${linkedId}.json`);
    await symlink(external, linked);
    assert.deepEqual(
      (await library.state()).entries.map((entry) => entry.id),
      [saved.id],
    );
    await assert.rejects(library.delete([saved.id, unknownId]));
    assert.equal((await library.load(saved.id)).artifact.kind, 'prepared');
    await assert.rejects(library.load('../external'));
    await assert.rejects(library.delete([linkedId]));
    await assert.rejects(library.delete(['../../notes.json']));
    assert.deepEqual((await library.delete([saved.id, saved.id])).entries, []);
    assert.equal(await readFile(ordinary, 'utf8'), '{"important":"keep"}');
    assert.equal(await readFile(unknown, 'utf8'), '{"unrelated":true}');
    assert.equal(await readFile(external, 'utf8'), 'external content');
    assert.equal((await readdir(options.defaultDirectory)).length, 3);
  } finally {
    await cleanup();
  }
});

test('tampered managed documents are neither loaded, overwritten nor deleted', async () => {
  const { library, options, cleanup } = await fixture();
  try {
    const prepared = await prepareRequest(miraRequest());
    const entry = await library.save(prepared);
    const filename = join(options.defaultDirectory, `unitlab-${entry.id}.json`);
    const document = JSON.parse(await readFile(filename, 'utf8'));
    document.id = '2'.repeat(64);
    const tampered = JSON.stringify(document);
    await writeFile(filename, tampered);
    assert.deepEqual((await library.state()).entries, []);
    await assert.rejects(library.load(entry.id), /identifier/);
    await assert.rejects(library.save(prepared), /identifier/);
    await assert.rejects(library.delete([entry.id]), /identifier/);
    assert.equal(await readFile(filename, 'utf8'), tampered);
  } finally {
    await cleanup();
  }
});

test('library HTTP endpoints require the session and local mutation origin', async () => {
  const { root, library, cleanup } = await fixture();
  const lab = await startLab({ library, model: new FakeModel(), example: miraRequest(), port: 0 });
  const headers = {
    Authorization: `Bearer ${lab.token}`,
    Origin: lab.origin,
    'Content-Type': 'application/json',
  };
  const post = (operation: string, body: unknown, override = {}) =>
    fetch(`${lab.origin}/api/library/${operation}`, {
      method: 'POST',
      headers: { ...headers, ...override },
      body: JSON.stringify(body),
    });
  try {
    assert.equal((await fetch(`${lab.origin}/api/library`)).status, 401);
    assert.equal(
      (await post('configure', { directory: join(root, 'new') }, { Origin: '' })).status,
      403,
    );
    const prepared = await prepareRequest(miraRequest());
    const saved = await post('save', { artifact: prepared });
    assert.equal(saved.status, 200);
    const entry = (await saved.json()) as LibraryEntry;
    const listing = await fetch(`${lab.origin}/api/library`, { headers });
    assert.deepEqual(((await listing.json()) as LibraryState).entries, [entry]);
    const loaded = await post('load', { id: entry.id });
    assert.deepEqual(await loaded.json(), { artifact: prepared });
    const removed = await post('delete', { ids: [entry.id] });
    assert.deepEqual(((await removed.json()) as LibraryState).entries, []);
    const configured = await post('configure', { directory: join(root, 'new') });
    assert.equal(((await configured.json()) as LibraryState).directory, join(root, 'new'));
    assert.equal((await post('save', { artifact: prepared, apiKey: 'reject-extra' })).status, 400);
    const draft = await draftUnit(prepared, new FakeModel());
    assert.equal((await post('icons', { artifact: draft }, { Authorization: '' })).status, 401);
    assert.equal((await post('icons', { artifact: draft }, { Origin: '' })).status, 403);
    const icons = await post('icons', { artifact: draft });
    assert.equal(icons.status, 200);
    assert.equal(((await icons.json()) as LibraryIconsResponse).icons.length, 17);
  } finally {
    await lab.close();
    await cleanup();
  }
});

test('icon paths are automatic, contained and stable across stages and revisions', async () => {
  const { library, options, cleanup } = await fixture();
  try {
    const prepared = await prepareRequest(miraRequest());
    assert.deepEqual(await library.icons(prepared), {
      directory: options.defaultDirectory,
      icons: [],
    });
    const candidate = miraCandidate();
    candidate.abilities.push({
      ...candidate.abilities[0]!,
      id: '../../form:active',
      pathId: null,
      tier: null,
    });
    const model = new FakeModel([candidate]);
    const draft = await draftUnit(prepared, model);
    const first = await library.icons(draft);
    assert.equal(first.icons.length, 18);
    assert.equal(first.icons[0]?.key, 'unit-portrait');
    assert.ok(first.icons.some((icon) => icon.key === 'tier:perception:2'));
    assert.ok(first.icons.some((icon) => icon.key === 'ability:../../form:active'));
    assert.ok(!first.icons.some((icon) => icon.key === 'ability:wall-perception'));
    for (const icon of first.icons) {
      assert.equal(dirname(icon.path), first.directory);
      assert.match(icon.path, /icon-[a-f0-9]{64}\.png$/);
      assert.equal(icon.dataUrl, undefined);
      assert.equal(icon.note, undefined);
    }
    assert.deepEqual(await library.icons(await checkDraft(draft)), first);
    const revision = structuredClone(draft);
    revision.run.id = 'later-run';
    revision.candidate.role = 'Revised role.';
    assert.deepEqual(await library.icons(revision), first);
    assert.deepEqual(await (await LabLibrary.open(options)).icons(revision), first);
    assert.equal((await library.state()).entries.length, 0);
  } finally {
    await cleanup();
  }
});

test('icon previews read existing PNGs with individual errors and bounded output', async () => {
  const { root, library, cleanup } = await fixture();
  try {
    const draft = await draftUnit(await prepareRequest(miraRequest()), new FakeModel());
    const first = await library.icons(draft);
    const png = Buffer.from(
      'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/x8AAwMCAO+a7VQAAAAASUVORK5CYII=',
      'base64',
    );
    await writeFile(first.icons[0]!.path, png);
    await writeFile(first.icons[1]!.path, 'not an image');
    const external = join(root, 'private.png');
    await writeFile(external, png);
    await symlink(external, first.icons[2]!.path);
    await writeFile(first.icons[3]!.path, Buffer.alloc(8 * 1024 * 1024 + 1));
    const loaded = await library.icons(draft);
    assert.equal(loaded.icons[0]!.dataUrl, `data:image/png;base64,${png.toString('base64')}`);
    assert.match(loaded.icons[1]!.note!, /PNG/);
    assert.equal(loaded.icons[1]!.dataUrl, undefined);
    assert.match(loaded.icons[2]!.note!, /regular local PNG/);
    assert.equal(loaded.icons[2]!.dataUrl, undefined);
    assert.match(loaded.icons[3]!.note!, /8 MB/);
    assert.equal(loaded.icons[3]!.dataUrl, undefined);
    const large = Buffer.alloc(2 * 1024 * 1024);
    png.copy(large);
    for (const icon of first.icons.slice(4)) await writeFile(icon.path, large);
    const bounded = await library.icons(draft);
    assert.ok(bounded.icons.some((icon) => icon.note?.includes('combined 16 MB')));
    assert.ok(
      bounded.icons.reduce((total, icon) => total + (icon.dataUrl?.length ?? 0), 0) <=
        16 * 1024 * 1024,
    );
    const entry = await library.save(draft);
    await library.delete([entry.id]);
    assert.deepEqual(await readFile(first.icons[0]!.path), png);
    assert.deepEqual(await readFile(external), png);
  } finally {
    await cleanup();
  }
});

test('icon provisioning rejects symlinked asset folders', async () => {
  const { root, library, options, cleanup } = await fixture();
  try {
    const external = join(root, 'outside');
    await mkdir(external);
    await mkdir(options.defaultDirectory);
    await symlink(external, join(options.defaultDirectory, 'assets'));
    const draft = await draftUnit(await prepareRequest(miraRequest()), new FakeModel());
    await assert.rejects(library.icons(draft), /real directories/);
    assert.deepEqual(await readdir(external), []);
  } finally {
    await cleanup();
  }
});

test('saved Profiles live beneath the library, are validated, and never replace bundled ones', async () => {
  const { library, options, cleanup } = await fixture();
  try {
    const initial = await library.profiles();
    assert.deepEqual(
      initial.profiles.map(({ profile, builtIn }) => [profile.id, builtIn]),
      bundledProfiles.map((profile) => [profile.id, true]),
    );
    const copy = {
      ...structuredClone(defaultUnitProfile),
      id: 'quick-copy',
      name: 'Quick copy',
      rules: { ...structuredClone(defaultUnitProfile.rules), id: 'profile:quick-copy' },
    };
    const saved = await library.saveProfile(copy);
    assert.deepEqual(saved.profiles.at(-1), { profile: copy, builtIn: false });
    const file = join(options.defaultDirectory, 'profiles', 'quick-copy.json');
    assert.deepEqual(JSON.parse(await readFile(file, 'utf8')), copy);
    const renamed = { ...copy, name: 'Quick copy, edited' };
    assert.equal((await library.saveProfile(renamed)).profiles.at(-1)?.profile.name, renamed.name);

    await assert.rejects(library.saveProfile(structuredClone(defaultUnitProfile)), /read-only/);
    await assert.rejects(
      library.saveProfile({ ...copy, id: 'other-id' }),
      /rules document must have the ID profile:other-id/,
    );
    const { mechanicsDefinition: _removed, ...invalid } = copy;
    await assert.rejects(library.saveProfile(invalid));
    await assert.rejects(library.deleteProfile(defaultUnitProfile.id), /cannot be deleted/);

    await writeFile(join(options.defaultDirectory, 'profiles', 'broken.json'), '{');
    await symlink(file, join(options.defaultDirectory, 'profiles', 'linked.json'));
    assert.equal((await library.profiles()).profiles.length, bundledProfiles.length + 1);

    const afterDelete = await library.deleteProfile('quick-copy');
    assert.equal(afterDelete.profiles.length, bundledProfiles.length);
    await assert.rejects(library.deleteProfile('quick-copy'));
  } finally {
    await cleanup();
  }
});

File: src/cli/internal/library/library_test.go
package library

import (
	"bytes"
	"encoding/base64"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"testing"

	"github.com/mardwerk/unit-generator/src/cli/internal/parity"
	s "github.com/mardwerk/unit-generator/src/cli/internal/schema"
	"github.com/mardwerk/unit-generator/src/cli/internal/unit"
)

// stages returns a prepared request, draft, checked draft and Result of one
// recorded planned-route review.
func stages(t *testing.T) []any {
	t.Helper()
	entries, err := parity.Entries("reviewDraft")
	if err != nil {
		t.Fatal(err)
	}
	for _, entry := range entries {
		result, ok := parity.Output(entry)
		checked := parity.Arg(entry, 0).(*s.Object)
		draft, _ := checked.Get("draft")
		candidate, _ := draft.(*s.Object).Get("candidate")
		if _, planned := candidate.(*s.Object).Get("blueprint"); !ok || !planned {
			continue
		}
		prepared, _ := draft.(*s.Object).Get("prepared")
		return []any{prepared, draft, checked, result}
	}
	t.Fatal("no recorded review")
	return nil
}

func open(t *testing.T) (*Library, string) {
	t.Helper()
	root := t.TempDir()
	library, err := Open(filepath.Join(root, "settings.json"), filepath.Join(root, "library"))
	if err != nil {
		t.Fatal(err)
	}
	return library, root
}

func TestLibrarySavesEveryStageOnceAndRestoresIt(t *testing.T) {
	library, root := open(t)
	state, err := library.State()
	if err != nil || len(state.Entries) != 0 || state.Directory != filepath.Join(root, "library") {
		t.Fatalf("state %+v %v", state, err)
	}
	artifacts := stages(t)
	for _, artifact := range artifacts {
		before := s.Stringify(artifact)
		var wait sync.WaitGroup
		entries := make([]Entry, 2)
		for i := range entries {
			wait.Add(1)
			go func() {
				defer wait.Done()
				entries[i], err = library.Save(artifact)
			}()
		}
		wait.Wait()
		if err != nil || entries[0] != entries[1] {
			t.Fatalf("entries %+v %+v %v", entries[0], entries[1], err)
		}
		kind, _ := artifact.(*s.Object).Get("kind")
		if entries[0].Kind != kind || entries[0].Character.Name == "" {
			t.Errorf("entry %+v", entries[0])
		}
		loaded, err := library.Load(entries[0].ID)
		// Saving keeps the content; keys follow the contract order.
		if err != nil || parity.Canonical(loaded) != parity.Canonical(artifact) || s.Stringify(artifact) != before {
			t.Errorf("load changed the %v artifact: %v", kind, err)
		}
	}
	reopened, _ := Open(filepath.Join(root, "settings.json"), filepath.Join(root, "library"))
	first, _ := library.State()
	second, _ := reopened.State()
	if s.Stringify(s.FromGoValue(first)) != s.Stringify(s.FromGoValue(second)) || len(first.Entries) != 4 {
		t.Errorf("reopened %d entries", len(second.Entries))
	}
	files, _ := filepath.Glob(filepath.Join(root, "library", "unitlab-*"))
	json, markdown := 0, 0
	for _, file := range files {
		switch filepath.Ext(file) {
		case ".json":
			json++
		case ".md":
			markdown++
			content, _ := os.ReadFile(file)
			if !strings.HasPrefix(string(content), "# ") {
				t.Errorf("%s is not a render", file)
			}
		}
	}
	if json != 4 || markdown != 3 {
		t.Errorf("%d records, %d renders", json, markdown)
	}
	extra := s.Clone(artifacts[0]).(*s.Object).Set("apiKey", "must-not-save")
	if _, err := library.Save(extra); err == nil {
		t.Error("saved an unknown field")
	}
	tampered := s.Clone(artifacts[0]).(*s.Object)
	request, _ := tampered.Get("request")
	request.(*s.Object).Set("task", "Changed without preparation.")
	if _, err := library.Save(tampered); err == nil || !strings.Contains(err.Error(), "hash") {
		t.Errorf("tampered: %v", err)
	}
	if _, err := library.Save(request); err == nil || !strings.Contains(err.Error(), "completed stage") {
		t.Errorf("request: %v", err)
	}
}

func TestConfigurePersistsAndLeavesTheOldLibrary(t *testing.T) {
	library, root := open(t)
	prepared := stages(t)[0]
	saved, err := library.Save(prepared)
	if err != nil {
		t.Fatal(err)
	}
	alternate := filepath.Join(root, "chosen", "units")
	state, err := library.Configure(alternate)
	if err != nil || state.Directory != alternate || len(state.Entries) != 0 {
		t.Fatalf("configured %+v %v", state, err)
	}
	reopened, _ := Open(filepath.Join(root, "settings.json"), filepath.Join(root, "library"))
	if reopened.Directory() != alternate {
		t.Errorf("reopened at %s", reopened.Directory())
	}
	if _, err := os.Stat(filepath.Join(root, "library", "unitlab-"+saved.ID+".json")); err != nil {
		t.Error("the old library changed")
	}
	_, _ = library.Save(prepared)
	if state, _ := library.Configure(filepath.Join(root, "library")); len(state.Entries) != 1 {
		t.Errorf("entries %d", len(state.Entries))
	}
	if _, err := library.Configure("  "); err == nil || library.Directory() != filepath.Join(root, "library") {
		t.Error("configured a blank folder")
	}
}

func TestDeleteTouchesOnlyIntactManagedDocuments(t *testing.T) {
	library, root := open(t)
	directory := filepath.Join(root, "library")
	saved, _ := library.Save(stages(t)[0])
	ordinary := filepath.Join(directory, "notes.json")
	_ = os.WriteFile(ordinary, []byte(`{"important":"keep"}`), 0o600)
	unknownID := strings.Repeat("0", 64)
	_ = os.WriteFile(filepath.Join(directory, "unitlab-"+unknownID+".json"), []byte(`{"unrelated":true}`), 0o600)
	external := filepath.Join(root, "external.json")
	_ = os.WriteFile(external, []byte("external content"), 0o600)
	linkedID := strings.Repeat("1", 64)
	_ = os.Symlink(external, filepath.Join(directory, "unitlab-"+linkedID+".json"))
	if state, _ := library.State(); len(state.Entries) != 1 || state.Entries[0].ID != saved.ID {
		t.Fatalf("state %+v", state)
	}
	for _, ids := range [][]string{{saved.ID, unknownID}, {linkedID}, {"../../notes.json"}} {
		if _, err := library.Delete(ids); err == nil {
			t.Errorf("deleted %v", ids)
		}
	}
	if _, err := library.Load("../external"); err == nil {
		t.Error("loaded outside the library")
	}
	if state, err := library.Delete([]string{saved.ID, saved.ID}); err != nil || len(state.Entries) != 0 {
		t.Errorf("delete %v", err)
	}
	for path, want := range map[string]string{ordinary: `{"important":"keep"}`, external: "external content"} {
		if content, _ := os.ReadFile(path); string(content) != want {
			t.Errorf("%s changed", path)
		}
	}
}

func TestTamperedDocumentsAreNeitherLoadedNorReplaced(t *testing.T) {
	library, root := open(t)
	prepared := stages(t)[0]
	entry, _ := library.Save(prepared)
	path := filepath.Join(root, "library", "unitlab-"+entry.ID+".json")
	content, _ := os.ReadFile(path)
	tampered := strings.Replace(string(content), `"id": "`+entry.ID+`"`, `"id": "`+strings.Repeat("2", 64)+`"`, 1)
	_ = os.WriteFile(path, []byte(tampered), 0o600)
	if state, _ := library.State(); len(state.Entries) != 0 {
		t.Error("listed a tampered document")
	}
	if _, err := library.Load(entry.ID); err == nil {
		t.Error("loaded a tampered document")
	}
	if _, err := library.Save(prepared); err == nil {
		t.Error("replaced a tampered document")
	}
	if _, err := library.Delete([]string{entry.ID}); err == nil {
		t.Error("deleted a tampered document")
	}
	if now, _ := os.ReadFile(path); string(now) != tampered {
		t.Error("the tampered file changed")
	}
}

var onePixel, _ = base64.StdEncoding.DecodeString("iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/x8AAwMCAO+a7VQAAAAASUVORK5CYII=")

func TestIconPathsAreContainedStableAndBounded(t *testing.T) {
	library, root := open(t)
	artifacts := stages(t)
	if icons, err := library.Icons(artifacts[0]); err != nil || len(icons.Icons) != 0 {
		t.Fatalf("prepared icons %+v %v", icons, err)
	}
	first, err := library.Icons(artifacts[1])
	if err != nil {
		t.Fatal(err)
	}
	if first.Icons[0].Key != "unit-portrait" || first.Icons[0].Kind != "portrait" || !strings.Contains(first.Icons[0].ImagePrompt, "character portrait") || !strings.Contains(first.Icons[1].CodexPrompt, first.Icons[1].Path) {
		t.Errorf("first icon %+v", first.Icons[0])
	}
	for _, icon := range first.Icons {
		if filepath.Dir(icon.Path) != first.Directory || !strings.HasSuffix(icon.Path, ".png") || icon.DataURL != "" || icon.Note != "" {
			t.Errorf("icon %+v", icon)
		}
	}
	for _, later := range artifacts[2:] {
		if again, _ := library.Icons(later); s.Stringify(s.FromGoValue(again.Icons[0].Path)) != s.Stringify(s.FromGoValue(first.Icons[0].Path)) {
			t.Error("icon paths changed across stages")
		}
	}
	_ = os.WriteFile(first.Icons[0].Path, onePixel, 0o600)
	_ = os.WriteFile(first.Icons[1].Path, []byte("not an image"), 0o600)
	external := filepath.Join(root, "private.png")
	_ = os.WriteFile(external, onePixel, 0o600)
	_ = os.Symlink(external, first.Icons[2].Path)
	_ = os.WriteFile(first.Icons[3].Path, make([]byte, 8*1024*1024+1), 0o600)
	loaded, _ := library.Icons(artifacts[1])
	if loaded.Icons[0].DataURL != "data:image/png;base64,"+base64.StdEncoding.EncodeToString(onePixel) ||
		!strings.Contains(loaded.Icons[1].Note, "PNG") || !strings.Contains(loaded.Icons[2].Note, "regular local PNG") || !strings.Contains(loaded.Icons[3].Note, "8 MB") {
		t.Errorf("notes %q %q %q", loaded.Icons[1].Note, loaded.Icons[2].Note, loaded.Icons[3].Note)
	}
	large := make([]byte, 2*1024*1024)
	copy(large, onePixel)
	for _, icon := range first.Icons[4:] {
		_ = os.WriteFile(icon.Path, large, 0o600)
	}
	bounded, _ := library.Icons(artifacts[1])
	total, limited := 0, false
	for _, icon := range bounded.Icons {
		total += len(icon.DataURL)
		limited = limited || strings.Contains(icon.Note, "combined 16 MB")
	}
	if !limited || total > 16*1024*1024 {
		t.Errorf("total %d limited %v", total, limited)
	}
	entry, _ := library.Save(artifacts[1])
	_, _ = library.Delete([]string{entry.ID})
	if content, _ := os.ReadFile(first.Icons[0].Path); !bytes.Equal(content, onePixel) {
		t.Error("deleting a record removed its icons")
	}
}

func TestIconsRejectLinkedAssetFolders(t *testing.T) {
	library, root := open(t)
	external := filepath.Join(root, "outside")
	_ = os.Mkdir(external, 0o755)
	_ = os.MkdirAll(filepath.Join(root, "library"), 0o755)
	_ = os.Symlink(external, filepath.Join(root, "library", "assets"))
	if _, err := library.Icons(stages(t)[1]); err == nil || !strings.Contains(err.Error(), "real directories") {
		t.Errorf("err %v", err)
	}
	if entries, _ := os.ReadDir(external); len(entries) != 0 {
		t.Error("wrote through the link")
	}
}

func TestSaveIconKeepsReceipts(t *testing.T) {
	library, _ := open(t)
	draft := stages(t)[1]
	cost := 0.004
	icons, err := library.SaveIcon(draft, "basic-attack", onePixel, Receipt{Model: "meta/muse-image", Usage: &unit.Usage{CostUSD: &cost}})
	if err != nil {
		t.Fatal(err)
	}
	if icons.Icons[1].DataURL == "" {
		t.Error("the saved icon is not shown")
	}
	receipts, _ := filepath.Glob(filepath.Join(icons.Directory, "image-*.json"))
	if len(receipts) != 1 {
		t.Errorf("receipts %d", len(receipts))
	}
	if _, err := library.SaveIcon(draft, "../escape", onePixel, Receipt{}); err == nil {
		t.Error("saved an icon without a destination")
	}
	if _, err := library.SaveIcon(draft, "basic-attack", []byte("not a png"), Receipt{}); err == nil {
		t.Error("saved a non-PNG")
	}
}

func TestProfilesAreValidatedAndNeverReplaceBundledOnes(t *testing.T) {
	directory := filepath.Join(t.TempDir(), "profiles")
	profiles, _ := OpenProfiles(directory)
	state, err := profiles.List()
	if err != nil || len(state.Profiles) != 1 || !state.Profiles[0].BuiltIn {
		t.Fatalf("state %+v %v", state, err)
	}
	copy := unit.DefaultProfile()
	copy.ID, copy.Name, copy.Rules.ID = "quick-copy", "Quick copy", "profile:quick-copy"
	saved, err := profiles.Save(s.FromGoValue(copy))
	if err != nil || len(saved.Profiles) != 2 || saved.Profiles[1].Profile.ID != "quick-copy" || saved.Profiles[1].BuiltIn {
		t.Fatalf("saved %v", err)
	}
	if content, _ := os.ReadFile(filepath.Join(directory, "quick-copy.json")); !strings.Contains(string(content), `"id": "quick-copy"`) {
		t.Error("profile file")
	}
	if got, err := profiles.Get("quick-copy"); err != nil || got.Name != "Quick copy" {
		t.Errorf("get %v", err)
	}
	if _, err := profiles.Save(s.FromGoValue(unit.DefaultProfile())); err == nil || !strings.Contains(err.Error(), "read-only") {
		t.Errorf("bundled: %v", err)
	}
	other := copy
	other.ID = "other-id"
	if _, err := profiles.Save(s.FromGoValue(other)); err == nil || !strings.Contains(err.Error(), "profile:other-id") {
		t.Errorf("other: %v", err)
	}
	invalid := s.FromGoValue(copy).(*s.Object)
	invalid.Delete("mechanicsDefinition")
	if _, err := profiles.Save(invalid); err == nil {
		t.Error("saved without a Definition")
	}
	if _, err := profiles.Delete(unit.DefaultProfile().ID); err == nil || !strings.Contains(err.Error(), "cannot be deleted") {
		t.Errorf("delete bundled: %v", err)
	}
	_ = os.WriteFile(filepath.Join(directory, "broken.json"), []byte("{"), 0o600)
	_ = os.Symlink(filepath.Join(directory, "quick-copy.json"), filepath.Join(directory, "linked.json"))
	if state, _ := profiles.List(); len(state.Profiles) != 2 {
		t.Errorf("listed %d", len(state.Profiles))
	}
	if state, err := profiles.Delete("quick-copy"); err != nil || len(state.Profiles) != 1 {
		t.Errorf("delete %v", err)
	}
	if _, err := profiles.Delete("quick-copy"); err == nil {
		t.Error("deleted twice")
	}
}
